README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
[![Travis-CI Build Status](https://travis-ci.org/ropenscilabs/rnaturalearth.svg?branch=master)](https://travis-ci.org/ropenscilabs/rnaturalearth)

[![Build status](https://ci.appveyor.com/api/projects/status/yp26qgeb1iligrpp?svg=true)](https://ci.appveyor.com/project/AndySouth/rnaturalearth)

rnaturalearth
=============

An R package to hold and facilitate interaction with [Natural Earth](http://www.naturalearthdata.com/) map data.

### Provides :

1.  access to a pre-downloaded subset of Natural Earth vector data commonly used in world mapping
2.  easy subsetting by countries and regions
3.  functions to download other Natural Earth vector and raster data
4.  a simple, reproducible and sustainable workflow from Natural Earth data to rnaturalearth enabling updating as new versions become available
5.  clarification of differences in world maps classified by countries, sovereign states and map units
6.  consistency with Natural Earth naming conventions so that rnaturalearth users can use Natural Earth documentation
7.  data in 'sf' or 'sp' formats

The [Natural Earth](http://www.naturalearthdata.com/) website structures vector data by scale, category and type. These determine the filenames of downloads. rnaturalearth uses this structure to facilitate download (like an API).

### Install rnaturalearth

You can install the development version from GitHub using [devtools](https://github.com/hadley/devtools).

    devtools::install_github("ropenscilabs/rnaturalearth")

Data to support much of the package functionality are stored in two data packages that you will be prompted to install when required if you do not do so here.

    devtools::install_github("ropenscilabs/rnaturalearthdata")
    install.packages("rnaturalearthhires",
                     repos = "http://packages.ropensci.org",
                     type = "source")

### First Usage

Here using `sp::plot` as a simple, quick way to plot maps. Maps could also be made with `ggplot2`, `tmap` or other options. All retrieval functions accept an argument `returnclass='sf'` to return package `sf` (Simple Features) objects.

``` r
library(rnaturalearth)
library(sp)

#world countries
sp::plot(ne_countries())
```

![](tools/README-unnamed-chunk-2-1.png)

``` r
#uk
sp::plot(ne_countries(country = 'united kingdom'))
```

![](tools/README-unnamed-chunk-2-2.png)

``` r
#states, admin level1 boundaries
sp::plot(ne_states(country = 'spain')) 
```

![](tools/README-unnamed-chunk-2-3.png)

### Introductory vignette

``` r
vignette('rnaturalearth', package='rnaturalearth')
```

### To download Natural Earth data not already in the package

There are a wealth of other data available at the [Natural Earth](http://www.naturalearthdata.com/) website. `rnaturalearth` has functions to help with download of these data.

The data available are outlined in the two tables below and online [here](http://www.naturalearthdata.com/downloads/50m-physical-vectors/).


    category   cultural 
                                    type scale110 scale50 scale10
    1                          countries     TRUE    TRUE    TRUE
    2                          map_units     TRUE    TRUE    TRUE
    3                       map_subunits    FALSE    TRUE    TRUE
    4                        sovereignty     TRUE    TRUE    TRUE
    5                     tiny_countries     TRUE    TRUE    TRUE
    6                             states    FALSE    TRUE    TRUE
    7                   populated_places     TRUE    TRUE    TRUE
    8                boundary_lines_land     TRUE    TRUE    TRUE
    9                  pacific_groupings     TRUE    TRUE    TRUE
    10          breakaway_disputed_areas    FALSE    TRUE    TRUE
    11     boundary_lines_disputed_areas    FALSE    TRUE    TRUE
    12 boundary_lines_maritime_indicator    FALSE    TRUE    TRUE
    13                          airports    FALSE    TRUE    TRUE
    14                             ports    FALSE    TRUE    TRUE
    15                       urban_areas    FALSE    TRUE    TRUE
    16                             roads    FALSE   FALSE    TRUE
    17                         railroads    FALSE   FALSE    TRUE

    category   physical 
                              type scale110 scale50 scale10
    1                    coastline     TRUE    TRUE    TRUE
    2                         land     TRUE    TRUE    TRUE
    3                        ocean     TRUE    TRUE    TRUE
    4      rivers_lake_centerlines     TRUE    TRUE    TRUE
    5                        lakes     TRUE    TRUE    TRUE
    6              glaciated_areas     TRUE    TRUE    TRUE
    7  antarctic_ice_shelves_polys     TRUE    TRUE    TRUE
    8             geographic_lines     TRUE    TRUE    TRUE
    9                 graticules_1     TRUE    TRUE    TRUE
    10                graticules_5     TRUE    TRUE    TRUE
    11               graticules_10     TRUE    TRUE    TRUE
    12               graticules_15     TRUE    TRUE    TRUE
    13               graticules_20     TRUE    TRUE    TRUE
    14               graticules_30     TRUE    TRUE    TRUE
    15          wgs84_bounding_box     TRUE    TRUE    TRUE
    16                      playas    FALSE    TRUE    TRUE
    17               minor_islands    FALSE   FALSE    TRUE
    18                       reefs    FALSE   FALSE    TRUE

Specify the `scale`, `category` and `type` of the vector you want as in the examples below.

``` r
#lakes
lakes110 <- ne_download(scale = 110, type = 'lakes', category = 'physical')
sp::plot(lakes110)

#rivers
rivers50 <- ne_download(scale = 50, type = 'rivers_lake_centerlines', category = 'physical')
sp::plot(rivers50)
```

### Details of different country definitions and scales

``` r
vignette('what-is-a-country', package='rnaturalearth')
```

Reproducible download of Natural Earth data into the package
------------------------------------------------------------

[Script](https://github.com/ropenscilabs/rnaturalearthdata/blob/master/data-raw/data_download_script.r) used to get data into the accompanying data packages.

Acknowledgements
----------------

Thanks to [Lincoln Mullen](https://github.com/lmullen) for code structure inspiration from [USAboundaries](https://github.com/ropensci/USAboundaries), [Hadley Wickham](https://github.com/hadley) for comments and prompting, [Bob Rudis](https://github.com/hrbrmstr) for answers to stackoverflow questions about downloading Natural Earth data into R. The [Natural Earth team](http://www.naturalearthdata.com/about/contributors/) and [Nathan Kelso](https://github.com/nvkelso) for providing such a great resource.

Potential future work
---------------------

### potential additional data

1.  Country synonyms lookup
    -   dataframe with ISO3 and country synonyms
    -   similar to <https://github.com/AndySouth/rworldmap/blob/master/data/countrySynonyms.rda>

2.  Country larger regions lookup
    -   dataframe with ISO3 and membership of different regional groupings, e.g. continent, least developed countries etc.
    -   similar to <https://github.com/AndySouth/rworldmap/blob/master/data/countryRegions.rda>

### potential additional functions

1.  facilitate joining of user data to country boundaries
    -   similar to <https://github.com/AndySouth/rworldmap/blob/master/R/joinCountryData2Map.R>
    -   ... but with a better name
    -   similar allowing of join by ISO codes or names, with attempted synonym matching
    -   similar reporting of country joining success and failure

2.  facilitate subsetting by country groupings
    -   e.g. least developed countries etc.

[![ropensci\_footer](http://ropensci.org/public_images/github_footer.png)](http://ropensci.org)


NEWS.md
rnaturalearth 0.1.0
===================

* Initial release
* sf support

R/ne_file_name.r
#' return a natural earth filename based on arguments
#'
#' returns a string that can then be used to download the file.
#' 
#' @param scale scale of map to return, one of \code{110}, \code{50}, \code{10} or \code{'small'}, \code{'medium'}, \code{'large'}
#' @param type type of natural earth file to download one of 'countries', 'map_units', 'map_subunits', 'sovereignty', 'states'
#'    OR the portion of any natural earth vector url after the scale and before the . 
#'    e.g. for 'ne_50m_urban_areas.zip' this would be 'urban_areas'
#'    OR the raster filename e.g. for 'MSR_50M.zip' this would be 'MSR_50M'
#' @param category one of natural earth categories : 'cultural', 'physical', 'raster'
#' @param full_url whether to return just the filename [default] or the full URL needed for download

#' @examples
#' ne_name <- ne_file_name( scale = 110, type = 'countries' )
#' ne_url  <- ne_file_name( scale = 110, type = 'countries', full_url = TRUE )
#' 
#' @return string
#' @export

ne_file_name <- function(scale = 110,
                    type = 'countries',
                    category = c('cultural', 'physical', 'raster'),
                    full_url = FALSE
) 
{
  
  # check on permitted scales, convert names to numeric
  scale <- check_scale(scale)
  
  # check permitted category
  category <- match.arg(category) 
  
  # check type against lists in package to warn user if they try something
  # that seems not to exist
  # however continue to try the download to give user flexibility
  # to get something I may have missed out of my list or has been added more recently
  check_data_exist( scale = scale, category = category, type = type )
  

  # add admin_0 to known types
  if ( type %in% c('countries', 'map_units', 'map_subunits', 'sovereignty', 'tiny_countries',
                   'boundary_lines_land', 
                   'pacific_groupings', 
                   'breakaway_disputed_areas',
                   'boundary_lines_disputed_areas',
                   'boundary_lines_maritime_indicator') )
  {
    type <- paste0('admin_0_',type)    
  }

  
  # add admin_1 to known types
  # this actually just expands 'states' to the name including lakes
  if (type == 'states')
    type <- 'admin_1_states_provinces_lakes'

 
  if (category=='raster')
  {
    # raster seems not to have so straightforward naming, so require that name is passed in type
    file_name <- paste0(type)
  } else {
    file_name <- paste0('ne_',scale,'m_',type)
  }

  
  if (full_url)
    file_name <- paste0('http://www.naturalearthdata.com/http//',
                      'www.naturalearthdata.com/download/',scale,'m/',category,'/',
                      file_name,'.zip' )
  
  return(file_name)
  
}